node/actors/bft/src/testonly/twins/scenario.rs

```diff
@@ -0,0 +1,182 @@
+
+        let twins = replicas
+            .iter()
+            .take(num_twins)
+            .map(|r| r.to_twin())
+            .collect::<Vec<_>>();
+
+        let num_replicas = replicas.len();
+
+        let mut nodes = replicas;
+        nodes.extend(twins);
+
+        Self {
+            nodes,
+            num_replicas,
+        }
+    }
+
+    /// The number of nodes in the cluster, both replicas and twins.
+    pub fn num_nodes(&self) -> usize {
+        self.nodes.len()
+    }
+
+    /// The number of original replicas, ie. not twins.
+    ///
+    /// This is the `n` in the `n = 5 * f + 1` equation.
+    pub fn num_replicas(&self) -> usize {
+        self.num_replicas
+    }
+
+    /// The number of twins in the cluster.
+    pub fn num_twins(&self) -> usize {
+        self.num_nodes() - self.num_replicas()
+    }
+
+    /// The number of faulty nodes the committee can tolerate for the `n = 5 * f + 1` relation to hold.
+    pub fn num_faulty(&self) -> usize {
+        (self.num_replicas - 1) / 5
+    }
+
+    /// The number of signatures required for a quorum, ie. `4 * f + 1`
+    pub fn quorum_size(&self) -> usize {
+        self.num_faulty() * 4 + 1
+    }
+
+    /// The number of signatures required for a high-vote, ie. `2 * f + 1`
+    pub fn subqourum_size(&self) -> usize {
+        self.num_faulty() * 2 + 1
+    }
+
+    /// All nodes in the cluster, replicas and twins.
+    pub fn nodes(&self) -> &[T] {
+        &self.nodes
+    }
+
+    /// The original replicas in the cluster.
+    pub fn replicas(&self) -> &[T] {
+        &self.nodes[..self.num_replicas]
+    }
+
+    /// The twins we created in the cluster.
+    pub fn twins(&self) -> &[T] {
+        &self.nodes[self.num_replicas..]
+    }
+}
+
+/// The configuration for a single round.
+pub struct RoundConfig<'a, T: HasKey> {
+    pub leader: &'a T::Key,
+    pub partitions: Partitioning<'a, T>,
+}
+
+/// Configuration for a number of rounds.
+pub struct Scenario<'a, T: HasKey> {
+    pub rounds: Vec<RoundConfig<'a, T>>,
+}
+
+pub struct ScenarioGenerator<'a, T>
+where
+    T: Twin,
+{
+    /// Number of rounds (views) to simulate in a scenario
+    num_rounds: usize,
+    /// Unique leader keys.
+    keys: Vec<&'a T::Key>,
+    /// All partitionings of various sizes we can choose in a round.
+    partitions: Vec<Partitioning<'a, T>>,
+}
+
+impl<'a, T> ScenarioGenerator<'a, T>
+where
+    T: Twin,
+{
+    /// Initialise a scenario generator from a cluster.
+    pub fn new(cluster: &'a Cluster<T>, num_rounds: usize, max_partitions: usize) -> Self {
+        assert!(!cluster.nodes().is_empty(), "empty cluster");
+
+        // Potential leaders
+        let keys = cluster
+            .replicas()
+            .iter()
+            .map(|r| r.key())
+            .collect::<Vec<_>>();
+
+        // Create all possible partitionings; the paper considers 2 or 3 partitions to be enough.
+        let partitions = (1..=max_partitions).flat_map(|np| partitions(cluster.nodes(), np));
```

Comment:
can you make the naming a bit more consistent? Right now `partitions` variable is actually partitionings which consist of partitions.

Reply:
Yeah, I know. It's just "partitionings" is a daft name 🤔

Reply:
Renamed to `splits`